File: judo-sdk/src/test/java/com/judopay/model/AddressTest.java
package com.judopay.model;

import org.junit.Test;

import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThat;

public class AddressTest {

    @Test
    public void shouldConstructAddressWithCorrectPropertyAssingment() {
        Address address = new Address("line1", "line2", "line3", "town", "postcode", 1);

        assertThat(address.getLine1(), equalTo("line1"));
        assertThat(address.getLine2(), equalTo("line2"));
        assertThat(address.getLine3(), equalTo("line3"));
        assertThat(address.getTown(), equalTo("town"));
        assertThat(address.getPostCode(), equalTo("postcode"));
        assertThat(address.getCountryCode(), equalTo(1));
    }

    @Test
    public void shouldBuildAddress() {
        Address address = new Address.Builder()
                .setLine1("line1")
                .setLine2("line2")
                .setLine3("line3")
                .setTown("town")
                .setPostCode("postcode")
                .setCountryCode(1)
                .build();

        assertThat(address.getLine1(), equalTo("line1"));
        assertThat(address.getLine2(), equalTo("line2"));
        assertThat(address.getLine3(), equalTo("line3"));
        assertThat(address.getTown(), equalTo("town"));
        assertThat(address.getPostCode(), equalTo("postcode"));
        assertThat(address.getCountryCode(), equalTo(1));
    }

    @Test
    public void shouldBuildPostCodeOnly() {
        Address address = new Address.Builder()
                .setPostCode("postcode")
                .build();

        assertThat(address.getLine1(), equalTo(null));
        assertThat(address.getLine2(), equalTo(null));
        assertThat(address.getLine3(), equalTo(null));
        assertThat(address.getTown(), equalTo(null));
        assertThat(address.getPostCode(), equalTo("postcode"));
        assertThat(address.getCountryCode(), equalTo(0));
    }
}

File: judokit-android/src/test/java/com/judopay/judokit/android/api/model/request/AddressTest.kt
package com.judopay.judokit.android.api.model.request

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("Testing Address.Builder")
internal class AddressTest {
    @Test
    @DisplayName("Given all fields are set, then build() returns correct Address")
    fun buildWithAllFields() {
        val address =
            Address
                .Builder()
                .setLine1("10 Downing Street")
                .setLine2("Westminster")
                .setLine3("Floor 1")
                .setTown("London")
                .setBillingCountry("GB")
                .setPostCode("SW1A 2AA")
                .setCountryCode(826)
                .setAdministrativeDivision("England")
                .build()

        assertEquals("10 Downing Street", address.line1)
        assertEquals("Westminster", address.line2)
        assertEquals("Floor 1", address.line3)
        assertEquals("London", address.town)
        assertEquals("GB", address.billingCountry)
        assertEquals("SW1A 2AA", address.postCode)
        assertEquals(826, address.countryCode)
        assertEquals("England", address.administrativeDivision)
    }

    @Test
    @DisplayName("Given no fields are set, then build() returns Address with all null fields")
    fun buildWithNoFields() {
        val address = Address.Builder().build()
        assertNull(address.line1)
        assertNull(address.line2)
        assertNull(address.line3)
        assertNull(address.town)
        assertNull(address.billingCountry)
        assertNull(address.postCode)
        assertNull(address.countryCode)
        assertNull(address.administrativeDivision)
    }

    @Test
    @DisplayName("Given blank administrativeDivision is set, then build() returns null for that field")
    fun buildWithBlankAdministrativeDivision() {
        val address =
            Address
                .Builder()
                .setAdministrativeDivision("   ")
                .build()
        assertNull(address.administrativeDivision)
    }

    @Test
    @DisplayName("Given empty administrativeDivision is set, then build() returns null for that field")
    fun buildWithEmptyAdministrativeDivision() {
        val address =
            Address
                .Builder()
                .setAdministrativeDivision("")
                .build()
        assertNull(address.administrativeDivision)
    }

    @Test
    @DisplayName("Given deprecated setState is called, then administrativeDivision is populated")
    fun setStatePopulatesAdministrativeDivision() {
        @Suppress("DEPRECATION")
        val address =
            Address
                .Builder()
                .setState("California")
                .build()
        assertEquals("California", address.administrativeDivision)
    }

    @Test
    @DisplayName("Given state getter is used, it returns same value as administrativeDivision")
    fun deprecatedStateGetterReturnsAdministrativeDivision() {
        val address =
            Address
                .Builder()
                .setAdministrativeDivision("Texas")
                .build()
        @Suppress("DEPRECATION")
        assertEquals("Texas", address.state)
    }
}
